[Joshua Yong] iP

jyrw opened the pull request:
# Duke Chatbot :rocket:
Here are the tasks in your list:
- [x] Go to GitHub
- [ ] Download Duke
- [ ] Manage your time better!

> Duke has helped me identify what needs to be done and manage my time better. Definitely recommend!
> -- A satisfied user

**Duke** is a *Personal Assistant Chatbot* designed for people who type fast. It features: 
* **Command-line interface** for keyboard-only navigation
* **Adding** and **removing** tasks
* Marking tasks as done
* **Filtering** tasks by name

With more features to come!

Using Duke is simple:
1. Click on [this link](https://github.com/jyrw/ip/releases/tag/A-Jar)
2. Download the `.jar` file
3. Double click the file
4. Start adding tasks!

Duke is a lightweight application written in Java. Here's the `main` method:
```Java
public static void main(String[] args) {
    new Duke("./data/duke.txt").start();
}
```

Commit:
Add Gradle support

jyrw committed:
implement skeletal version of Duke

jyrw committed:
implement add, list

jyrw committed:
add Task.java

jyrw committed:
implement marking tasks as done

jyrw committed:
add Deadline.java, Event.java, ToDo.java

jyrw committed:
remove getStatusIcon method, implement toString method

jyrw committed:
fix string formatting issues in toString method

jyrw committed:
implement Deadlines, Events, ToDos

jyrw committed:
refactor code, change task array to ArrayList

jyrw committed:
fix raw types

jyrw committed:
add input.txt, EXPECTED.txt

jyrw committed:
add DukeException.java

jyrw committed:
implement error handling with DukeException

jyrw committed:
update EXPECTED.txt

jyrw committed:
implement delete

jyrw committed:
update input.txt, EXPECTED.txt

jyrw committed:
fix DukeException.java

jyrw committed:
Update .gitignore

jyrw committed:
Implement saving task list to text file

jyrw committed:
Implement dates and times in 'deadline' and 'event' commands

jyrw committed:
Merge branch 'branch-Level-7'

jyrw committed:
Merge branch 'branch-Level-8'

jyrw committed:
Modify toString method

jyrw committed:
Refactor code to extract Ui, Storage, Parser, and TaskList classes

jyrw committed:
Organize classes into packages

jyrw committed:
Add JUnit tests for Task class

jyrw committed:
Add manifest

jyrw committed:
Add JavaDoc comments

jyrw committed:
Tweak code to comply with coding standards

jyrw committed:
Implement find

jyrw committed:
Merge branch 'branch-A-JavaDoc'

jyrw committed:
Merge branch 'branch-A-CodingStandard'

jyrw committed:
Merge branch 'branch-Level-9'

jyrw committed:
Merge branch 'add-gradle-support' of https://github.com/jyrw/ip into branch-A-Gradle

jyrw committed:
Merge branch 'branch-A-Gradle'

jyrw committed:
Add config files for Checkstyle

jyrw committed:
Fix style issues

jyrw committed:
Merge branch 'branch-A-CheckStyle'

Preshita01 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,123 @@
+     * @return A String to be displayed to the user.
+     * @throws DukeException if the user input is invalid.
+     */
+    public String parse(String userInput) throws DukeException {
```
Maybe you could consider breaking up this 1 long method into 2 methods? The first method could throw errors based on the user's input (if any). If there are no errors, the first method could then call on the second method to parse the input

Preshita01 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,76 @@
+        this.tasks = tasks;
+    }
+
+    public ArrayList<Task> getTasks() {
```
Perhaps you can consider modifying the names of one of the methods (`getTasks()` or `getTask(int i)`) so that they are more distinguishable?

Preshita01 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,76 @@
+        tasks.add(task);
+    }
+
+    public int getSize() {
```
Maybe you can name this as `getTasksSize()` or `numOfTasks()` so that it is clearer that 'size' here refers to size of `tasks`?

Preshita01 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,76 @@
+        tasks.get(i).setDone();
+    }
+
+    public void deleteTask(int i) {
```
Maybe you can include more descriptive header comments for the public non-setter and non-getter methods?

Preshita01 reviewed (commented):
The overall code quality seems good, with just a few points for improvement 👍

jyrw committed:
Implement GUI

jyrw committed:
Merge branch 'branch-Level-10'

hanif-kamal commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,123 @@
+     *
+     * @param userInput The String inputted by the user.
+     * @return A String to be displayed to the user.
+     * @throws DukeException if the user input is invalid.
```
Perhaps you could capitalise the first letter of the sentence after the word DukeException?

```suggestion
     * @throws DukeException If the user input is invalid.
```

hanif-kamal commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,123 @@
+ */
+public class Parser {
+
+    // The TaskList which the Parser updates
```
Perhaps you could change this comment about this class member to a single-line Javadoc like so: 
```suggestion
    /** The TaskList which the Parser updates */
```

hanif-kamal commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /**
+     * Parses the associated file, returning a collection of Tasks.
```
Maybe there should be an empty line between the description and parameter section? Like so:
```suggestion
     * Parses the associated file, returning a collection of Tasks.
     *
```

hanif-kamal commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,106 @@
+    /**
+     * Parses the associated file, returning a collection of Tasks.
+     * @return An ArrayList of Tasks corresponding to the data in the file.
+     * @throws DukeException if the file is formatted incorrectly or cannot be found
```
Perhaps you could capitalise the first letter of the sentence after the word DukeException? Same as above: 
```suggestion
     * @throws DukeException If the file is formatted incorrectly or cannot be found
```

hanif-kamal reviewed (commented):
Overall, really well done! Just some minor tweaks that you could consider.

jyrw committed:
Implement assertions

jyrw committed:
Improve code quality and fix coding standard violations

There are various code quality issues in the code.
An example is an 80 line method in the Parser class.
Another example is a redundant Ui class.
Additionally, there are numerous coding standard violations such as
missing Javadocs.

These issues make the code harder to read and decipher.
They will lead to the program being more difficult to debug in the
future.

Extracting out some logic from the offending method in the Parser class
into smaller methods helps to minimise the complexity of the code.
Writing Javadocs for more public methods also increases readability.

These changes help to make the codebase more maintainable.

jyrw committed:
Merge pull request #1 from jyrw/branch-A-CodeQuality

Improve code quality and fix coding standard violations

jyrw committed:
Merge branch 'master' of https://github.com/jyrw/ip

* 'master' of https://github.com/jyrw/ip:
  Improve code quality and fix coding standard violations

jyrw committed:
Merge branch 'master' into branch-A-Assertions

jyrw committed:
Merge pull request #2 from jyrw/branch-A-Assertions

Implement assertions

jyrw committed:
Implement showing upcoming deadlines on startup

jyrw committed:
Merge branch 'branch-B-Reminders'

jyrw committed:
Standardise comments

jyrw committed:
Fix typos and code quality issues

jyrw committed:
Add the final keyword to some attributes

jyrw committed:
Improve the GUI

jyrw committed:
Fix bug with reading Events from Storage text file

jyrw committed:
Move showing upcoming deadlines to dedicated command

jyrw committed:
Polish GUI further

jyrw committed:
Rename chatbot to Cygnus

jyrw committed:
Revert README.md

jyrw committed:
Update docs/README.md

jyrw committed:
Add image of UI

jyrw committed:
Organize GUI classes into package

jyrw committed:
Fix code quality issues

jyrw committed:
Implement help command

jyrw committed:
Update docs/README.md

jyrw committed:
Fix main class name in Gradle build file

jyrw committed:
Add JUnit tests

jyrw committed:
Fix JUnit tests

jyrw committed:
Fix style issues in JUnit tests

jyrw committed:
Fix UI image file name

jyrw committed:
Update README.md
